util/src/overlaydb.rs

```diff
@@ -40,340 +47,167 @@ use kvdb::{Database, DBTransaction};
-			true
-		}
+	/// Get the value of the given key.
+	fn payload(&self, key: &H256) -> Option<Bytes> {
+		// TODO [rob] have this and HashDB functions all return Results.
+		self.backing.get(key).expect("Low level database error.").map(|v| v.to_vec())
 	}
 }
 
 impl HashDB for OverlayDB {
 	fn keys(&self) -> HashMap<H256, i32> {
-		let mut ret: HashMap<H256, i32> = HashMap::new();
+		let mut ret = HashMap::new();
+
 		for (key, _) in self.backing.iter() {
-			let h = H256::from_slice(key.deref());
-			let r = self.payload(&h).unwrap().1;
-			ret.insert(h, r as i32);
+			ret.insert(H256::from_slice(&*key), 1);
 		}
 
-		for (key, refs) in self.overlay.keys().into_iter() {
-			let refs = *ret.get(&key).unwrap_or(&0) + refs;
-			ret.insert(key, refs);
+		for (key, refs) in self.overlay.keys() {
+			*ret.entry(key).or_insert(0) += refs;
 		}
+
 		ret
 	}
+
 	fn get(&self, key: &H256) -> Option<&[u8]> {
-		// return ok if positive; if negative, check backing - might be enough references there to make
-		// it positive again.
-		let k = self.overlay.raw(key);
-		match k {
-			Some(&(ref d, rc)) if rc > 0 => Some(d),
-			_ => {
-				let memrc = k.map_or(0, |&(_, rc)| rc);
-				match self.payload(key) {
-					Some(x) => {
-						let (d, rc) = x;
-						if rc as i32 + memrc > 0 {
-							Some(&self.overlay.denote(key, d).0)
-						}
-						else {
-							None
-						}
-					}
-					// Replace above match arm with this once https://github.com/rust-lang/rust/issues/15287 is done.
-					//Some((d, rc)) if rc + memrc > 0 => Some(d),
-					_ => None,
-				}
-			}
+		match self.overlay.raw(key) {
+			Some(&(ref d, rc)) if rc > 0 => { Some(d) }
+			_ => if let Some(x) = self.payload(key) {
+				Some(&self.overlay.denote(key, x).0)
+			} else {
+				None
+			},
 		}
 	}
+
 	fn contains(&self, key: &H256) -> bool {
-		// return ok if positive; if negative, check backing - might be enough references there to make
-		// it positive again.
-		let k = self.overlay.raw(key);
-		match k {
-			Some(&(_, rc)) if rc > 0 => true,
-			_ => {
-				let memrc = k.map_or(0, |&(_, rc)| rc);
-				match self.payload(key) {
-					Some(x) => {
-						let (_, rc) = x;
-						rc as i32 + memrc > 0
-					}
-					// Replace above match arm with this once https://github.com/rust-lang/rust/issues/15287 is done.
-					//Some((d, rc)) if rc + memrc > 0 => true,
-					_ => false,
-				}
-			}
-		}
+		self.get(key).is_some()
 	}
-	fn insert(&mut self, value: &[u8]) -> H256 { self.overlay.insert(value) }
-	fn emplace(&mut self, key: H256, value: Bytes) { self.overlay.emplace(key, value); }
-	fn remove(&mut self, key: &H256) { self.overlay.remove(key); }
-}
 
-#[test]
-fn overlaydb_overlay_insert_and_remove() {
-	let mut trie = OverlayDB::new_temp();
-	let h = trie.insert(b"hello world");
-	assert_eq!(trie.get(&h).unwrap(), b"hello world");
-	trie.remove(&h);
-	assert_eq!(trie.get(&h), None);
-}
+	fn insert(&mut self, value: &[u8]) -> H256 {
+		self.overlay.insert(value)
```

Comment:
Wouldn't it be better to have some assertions about refcount here? To detect errors earlier? (right know it might fail during `commit`)
Disregard if this comment makes no sense :)